lib/phoenix_live_view/async.ex

```diff
@@ -50,20 +57,36 @@ defmodule Phoenix.LiveView.Async do
     :ok
   end
 
+  defp validate_function_env(func, op) do
+    {:env, variables} = Function.info(func, :env)
+
+    if Enum.any?(variables, &match?(%Phoenix.LiveView.Socket{}, &1)) do
+      warn_socket_access(op, __ENV__)
```

Comment:
In this case, it may be better to pass nothing and warn based on the current stacktrace. So maybe, you pass the warning function as argument?

```suggestion
      warn_socket_access(op, fn msg, _meta -> IO.warn(msg) end)
```

And in the macro case you compose the meta with the `__ENV__` (and the Elixir version).